### DISCUSSION TITLE
ref field is marshaled as byte/sbyte, not pointer

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/runtime#93588] [General] [answered]
Test.cpp
```cpp
extern "C" __declspec(dllexport) void Test(int* pNums) {
	cout << (long long)pNums << endl;
}
```

Test.cs
```csharp
public ref struct RS
{
    public ref int pNums;
}

[DllImport("Test.cpp")]
public static extern void Test(RS rs);

[DllImport("Test.cpp")]
public static extern void Test(ref int pNums);

public static void Main()
{
    int[] nums = GC.AllocateArray<int>(10, true);

    Test(ref nums[0]);

    RS rs = new RS() { pNums = ref nums[0] };
    Test(rs);

    RS* pRS = &rs;
    sbyte* pB = (sbyte*)pRS;
    Console.WriteLine(*pB);
}
```

The method ```void Test(ref int pNums)``` will always output a pointer to ```nums[0]``` in console.
But the method ```void Test(RS rs)``` will alaways output 64 in console, not a pointer to ```nums[0]```.
At last, I discovered the ```RS rs``` will be marshaled as a byte/sbyte, not a pointer to ```nums[0]```.
Is this a bug?
![image](https://github.com/dotnet/runtime/assets/148173140/b50a208b-46d4-4d07-a6ce-1dc05a5dc829)

**Comment by huoyaoyuan** [comment]
Should ref field ever be supported in marshaling? @dotnet/interop-contrib

**Reply by AaronRobinsonMSFT** [reply]
> Should ref field ever be supported in marshaling?

If the target of the `ref` is blittable **or** `unmanaged` with [`DisableRuntimeMarshalling`](https://learn.microsoft.com/dotnet/standard/native-interop/disabled-marshalling) set, and the target of the `ref` is pinned, then yes marshalling could be performed.

Whether the above constraints can all be met safely is another matter.

@Maple1031 What version of the runtime is being used here?

**Reply by AaronRobinsonMSFT** [reply, marked as answer]
Note that the layout of the `RS` type is undefined since it contains a `managed` reference. Performing the pointer manipulation in the manner above is considered invalid.

**Reply by ghost** [reply]
>Note that the layout of the RS type is undefined since it contains a managed reference. Performing the pointer manipulation in the manner above is considered invalid.

Thanks.
